File: frontend/src/utils/passwordValidator.js
export const validatePassword = (password) => {
    if (!password || password.length < 8 || password.length > 32) return false;

    let qualifiers = 0;
    if (/[a-z]/.test(password)) qualifiers++;
    if (/[A-Z]/.test(password)) qualifiers++;
    if (/[0-9]/.test(password)) qualifiers++;
    if (/[^a-zA-Z0-9]/.test(password)) qualifiers++;

    return qualifiers >= 3;
};

File: backend/src/main/java/com/shepherdsstories/utils/PasswordValidator.java
package com.shepherdsstories.utils;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

import static com.shepherdsstories.utils.ValidationConstants.PASSWORD_MAX_LENGTH;
import static com.shepherdsstories.utils.ValidationConstants.PASSWORD_MIN_LENGTH;

public class PasswordValidator implements ConstraintValidator<ValidPassword, String> {

    @Override
    public boolean isValid(String password, ConstraintValidatorContext context) {
        if (password == null || password.length() < PASSWORD_MIN_LENGTH || password.length() > PASSWORD_MAX_LENGTH) {
            return false;
        }

        int qualifiers = 0;
        if (password.matches(".*[a-z].*")) qualifiers++;
        if (password.matches(".*[A-Z].*")) qualifiers++;
        if (password.matches(".*[0-9].*")) qualifiers++;
        if (password.matches(".*[^a-zA-Z0-9].*")) qualifiers++;

        return qualifiers >= 3;
    }
}
